Inventory-based SSH proxy/bastion host support, ssh_args + docs

nitzmahone opened the pull request:
Supersedes #9122 

First-class inventory var support for proxy/bastion/jump hosts seems to be increasingly requested. I've tried to make it reasonably comprehensive and well-documented, though I intentionally omitted proxy_password support (had some trouble stacking the file descriptors for the two-level-deep sshpass).

(CC @sivel)

Commit:
Made ansible_ssh_args runtime-settable, added proxy_host, port, user, and private_key_file support + docs

Only works in OpenSSH, added warning to paramiko if these values are encountered there.

jimsmith commented:
+1 for this, it's a request that I rarely get asked but it's becoming more frequently.

mpdehaan commented:
So this has been discussed a bit recently on the devel list, and I believe there may be some competing implementations, can you stop by ansible-devel to discuss?

Thanks!

tzz commented:
I've needed this so definitely happy to see an implementation!

quinot commented:
I'm just starting to use Ansible and was wondering whether this feature existed, or how to implement it if not. Would love to have this merged!

bcoca committed:
preliminary privlege escalation unification  + pbrun
- become constants inherit existing sudo/su ones
- become command line options, marked sudo/su as deprecated and moved sudo/su passwords to runas group
- changed method signatures as privlege escalation is collapsed to become
- added tests for su and become, diabled su for lack of support in local.py
- updated playbook,play and task objects to become
- added become to runner
- added whoami test for become/sudo/su
- added home override dir for plugins
- removed useless method from ask pass
- forced become pass to always be string also uses to_bytes
- fixed fakerunner for tests
- corrected reference in synchronize action plugin
- added pfexec (needs testing)
- removed unused sudo/su in runner init
- removed deprecated info
- updated pe tests to allow to run under sudo and not need root
- normalized become options into a funciton to avoid duplication and inconsistencies
- pushed suppored list to connection classs property
- updated all connection plugins to latest 'become' pe

- includes fixes from feedback (including typos)
- added draft docs
- stub of become_exe, leaving for future v2 fixes

rmetzler commented:
I really want this feature in Ansible, but after looking at the code, I would think you shouldn't add that many arguments to `connect`. Maybe use `*args` and `**kwargs`.

nitzmahone commented:
Funny, that was my first thought when I looked at it too- "why are they passing all these args positionally?" I looked at the history, and the last couple of args were added later. I assumed there was some reason to do with the abstraction of the Connect stuff, and went for the "when in Rome" approach, but if reviewers feel the same way, I'm happy to convert it to use named args instead. It _is_ much cleaner, IMHO.

rmetzler commented:
I want to say, I just said this as a user who needs this feature and searched through the different threads / PRs. I really don't have any authority regarding Ansible.

igoratencompass commented:
I've been waiting for this kind of module definitely needed for dynamic aws inventory of mixed public and private instances! Can we please have this released asap???

christophgysin commented:
I'm also looking forward to this!

christophgysin commented on lib/ansible/runner/connection_plugins/ssh.py:
```diff
@@ -58,10 +63,35 @@ def connect(self):
+                proxy_args += ["-o", "User="+self.proxy_user]
+            proxy_args += ["-o", "ConnectTimeout=%d" % self.runner.timeout]
+
+            self.common_args += ["-o", "ProxyCommand=ssh {proxy_host} {proxy_args} -W %h:%p".format(proxy_host=self.proxy_host, proxy_args=" ".join(proxy_args))]
```
ssh -W asumes openssh-5.4 or later. Older versions of openssh need netcat to achieve this.

nitzmahone replied:
Yeah, I considered using netcat as a lowest-common-denominator approach. However, I saw some complaints around the openssh community about instability with various versions of netcat that the -W mode attempts to address (don't have a reference to share off the top of my head- it was several months ago). If a bastion host can't be upgraded to a reasonably recent version of openssh, the ssh_args escape hatch is still there- I think that beats the alternatives of:
- forcing everyone to use a potentially unstable approach that has another dependency (netcat)
- adding yet another mode-changing variable (ansible_ssh_proxy_mode=netcat or some such blah)
- version-sniffing ssh (rife with issues)

anthcourtney commented:
Is there a timeline for this feature progressing / being integrated?

rmetzler commented:
For everyone needing a workaround fast, this is a great tutorial: http://alexbilbie.com/2014/07/using-ansible-with-a-bastion-host/

But I also would prefer first class support in Ansible.

mpeters commented:
That approach only works when you have a single bastion server for all your environments. This won't work if you have a separate bastion for say production, staging and integration.

bcoca commented:
@mpeters it does, you just need to specify the diff bastions per host or host pattern (this is where naming conventions can help a lot).

mpeters commented:
That's assuming you are naming your servers. How would you handle just IP addresses (since these are livestock, not pets) in different VPCs? The IPs can be repeated for a different environment. 

Also, how would you hook this up with Tower? (You'd need multiple Tower installs right?) Multiple Tower installs? 

What about a single environment that uses multiple VPCs for PCI, etc?

bcoca commented:
no need for multiple ansible/tower installs, one thing you can do is set ssh args per group/host/inventory, passing specific ssh config for each.

In any case, this is a workaround, the patch is a good step into a more integrated solution.

mpeters commented:
Can you really set ansible_ssh_args per group/host/inventory (that would be great)? The documentation doesn't list it with the rest of the ansible_ssh_\* args (http://docs.ansible.com/intro_inventory.html#list-of-behavioral-inventory-parameters) and the only place I see you can set it is in the ansible config file, which is global.

rmetzler commented:
I had a quick look into the source and it seems like `ssh_args` is respected in 1.8.2. Not sure since when, though.

https://github.com/ansible/ansible/blob/v1.8.2/lib/ansible/constants.py#L165
https://github.com/ansible/ansible/blob/v1.8.2/lib/ansible/runner/connection_plugins/ssh.py#L55-L64

mpeters commented:
But that just shows that it's configurable globally from the ansible configuration file. Not that it's possible to specify per group/host/inventory, right?

snackycracky commented:
I also need the 'Ansible EC2 external inventory script' to detect the internal hosts. http://docs.ansible.com/intro_dynamic_inventory.html#example-aws-ec2-external-inventory-script
it only finds the jump-hosts for each environment but not the internal servers.

bcoca committed:
removed bare variable detection as this confuses people and forced us to allow for bare expressions

bcoca committed:
changed from hash_merge to combine vars which resets default to
overwrite and not merge hashing
corrected merge vs combined in all pertinent sections
fixed typoe in combined_Vars
removed redundant inventory call, moved grousp to proper priority
readded inventory vars to runner's vars
correclty added inventory this time

htgoebel committed:
Fix detect of docker as virtualization_type.

Not only match`/docker/`, but also `docker-` followed by a hex-id.

Example (shortened):
```
$ cat /proc/1/cgroup
8:blkio:/system.slice/docker-de73f4d207861cf8757b69213ee67bb234b897a18bea7385964b6ed2d515da94.scope
7:net_cls:/
```

jimi-c committed:
Setting up new release candidate versioning

jimi-c committed:
Submodule update for stable-1.9 branch

jlaska committed:
Add tox and travis-ci support

Add tox integration to run unittests in supported python releases.
Travis-CI is used for test execution.

Additionally, the unittest TestQuotePgIdentifier was updated to support
using assert_raises_regexp on python-2.6.

Sample travis-ci output available at
https://travis-ci.org/ansible/ansible/builds/54189977

shirou committed:
use to_unicode() in _jinja2_vars if type is str.

jeffwidman committed:
Typo: lead --> led

bcoca committed:
fixed bad paren in connection plugin

bcoca committed:
fix issue with ask pass signature

bcoca committed:
fixed traceback when x_user implicitly sets the become method

Fixes #10430

Also removed redundant resolution of sudo/su for backwards compatibility which
confused the conflict detection code.

bcoca committed:
removed uneeded reference to su_user

jimi-c committed:
Fix deb packaging version in the changelog

bcoca committed:
fixes password error detection for ssh connection plugin
removes sycnronize test that does not work with current sudo setup
Fixes #10434

bcoca committed:
fix tag test that broke with new tag info displayed in list tasks

bcoca committed:
fixed  missed su to become conversion

jhermann committed:
Generic package_dir mapping in setup.py (closes #10437)

abadger committed:
Update core modules to pull in fixes

jhermann committed:
added test requirements for pip

abadger committed:
Update for another 3c2 fix

shirou committed:
set 'nonstring' arg to passthru.

abadger committed:
Test case for #10426

bcoca committed:
fixed missed conversion of su to become

abadger committed:
Update core pointer to pick up docker fix

jimi-c committed:
Fix issue with unarchive disabling pipelining mode

Was using persist_files=True when specifying the create paramater,
which breaks pipelining. Switched to use delete_remote_tmp=False instead,
which is the proper way to preserve the remove tmp dir when running
other modules from the action plugin.

bcoca committed:
fixed and reintroduced syncronize test, fakerunner object needed become_method to be it's default 'sudo'

abadger committed:
Hash randomization makes one of the heuristic_log_sanitize checks not work.

Nothing we can do, when it sanitizes ssh_urls it's simply overzealous.

abadger committed:
Comma is also dependent on position within the hash

jlaska committed:
Enable assert_raises_regexp on py26

abadger committed:
Update core module pointer

sgargan committed:
fix for issue #10422. outputs informative error message when AWS credentials are not available

bcoca committed:
fixed raw return check for privilege escalation

abadger committed:
Update core modules pointer

abadger committed:
Update core pointer to make use of DOCKER_TLS_VERIFY env var:

https://github.com/ansible/ansible-modules-core/issues/946

sgargan committed:
log errors and explicitly exit rather than raising exceptions

bcoca committed:
slight changes to allow for checksum and other commands to work correctly with quoting

bcoca committed:
fixed issue with su in plays

abadger committed:
Update core modules pointer

abadger committed:
Initial test of the docker module

abadger committed:
Ugh, looks like very few distros have the proper packages to run the docker module.

break up the tests so that we can maybe  run this on at least one
platform

abadger committed:
Would help if I added these files in the right directory

abadger committed:
Attempt to enable docker tests for rhel/centos6 as well

abadger committed:
And ran into a different problem with centos6.  Sigh.

abadger committed:
Add tests using a docker private registry

abadger committed:
Remove debug statements

bcoca committed:
added missing become method inventory override

abadger committed:
Add more tests for private docker registries

abadger committed:
Some debugging for why docker tests are failing in jenkins

abadger committed:
Fix the removal of busybox image

abadger committed:
Okay, let's see if these pauses are enough to get this passing

abadger committed:
Update docker module

abadger committed:
Have selinux allow docker<=>nginx communication

abadger committed:
Update core module pointer

abadger committed:
Update the extras module pointer

dopry commented:
@bcoca,  Are you comfortable with this approach to bastion host support, or is there a competing implementation you prefer? I'm currently working on a VPC deployment and can put some cycles in testing and getting the PR up to date if you guys are interested in getting it merged.

bcoca committed:
ignore PE methods that are not sudo for checksums until we get them working universally

bcoca committed:
updated core to latests stable 1.9

bcoca committed:
updated module ref for core

abadger committed:
Pull another fix in from core modules

bcoca committed:
now add_host loads hostvars

abadger committed:
Pull ec2_asg fixes by updating core modules

bcoca committed:
now correctly aplies add_host passed variables last to override existing vars.

bcoca committed:
removed debug play from tests

bcoca committed:
now use combine vars to preserve existing cached host vars

Commit:
Patch for bug #10485 - ansible_distribution fact populates as 'RedHat' on Oracle Linux systems

abadger committed:
Update core modules for asg tag fix

jimi-c committed:
VERSION bump and submodule update for 1.9.0-0.2.rc2

bcoca committed:
now ansible ignores tempate errors on passwords
they could be caused by random character combinations, fixes #10468

eriweb commented:
Thanks @nitzmahone, I'm testing this, but I seem to have trouble getting Ansible to wait for the proxycommand to finish.
I get this error message blazingly fast: "fatal: [r-4-VM] => SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh"
but if I try the same SSH command manually it works just fine..

bcoca committed:
makes raw module have quiet ssh so as to avoid extra output when not requried

bcoca committed:
updated to latest ref

abadger committed:
And all of core module changes added

jimi-c committed:
tweaking the CHANGELOG

abadger committed:
Fix assert to work with unicode values

abadger committed:
Make run_command() work when we get byte str with non-ascii characters (instead of unicode type like we were expecting)

Fix and test.

Fixes #10536

jimi-c committed:
Version bump for 1.9.0 final release

jimi-c committed:
cleaning up CHANGELOG with incorrect entries

jimi-c committed:
Updating submodule pointers for 1.9

jimi-c committed:
Submodule updates for 1.9

bcoca committed:
updated ref to extras

jimi-c committed:
Version bump for 1.9.0-2

jimi-c committed:
Version bump for 1.9.0.1-1

deimosfr committed:
fix consul inventory issue (missing method param)

bcoca committed:
updated sumbmodule refs

toscom-cru committed:
Update README.md

I think since ansible and the ansible-modules have been splitted --recursive should be added

bcoca committed:
Merge pull request #10556 from kristous/patch-1

Update README.md

bcoca committed:
removed folding sudo/su to become logic from constants as it is already present downstream in playbook/play/tasks

bcoca committed:
readded sudo/su vars to allow role/includes to work with passed sudo/su

bcoca committed:
added note that custom connection plugins need update with this version

bcoca committed:
dont break everything when one of the vars  in inject does not template correctly, wait till its used

bcoca committed:
converted error on play var initialization into warning with more information

bcoca committed:
capture IOErrors on backup_local (happens on non posix filesystems)

fixes #10591

bcoca committed:
updated submodule refs

bcoca commented:
I was thinking about integrating this soon now that v2 is almost 'there', but needs massive rebase, should probably wait till v2 is in place in devel

bcoca commented:
the other thing I wanted to do is move away from having ansiible_ssh and go for ansible_connection (or at least conn) as the variable naming, so we can actually use the same vars for other protocols w/o asking 'why does winrm use ansible_ssh_user?'

nitzmahone commented:
Yeah, this will probably just end up being a whole new PR for v2 (adding support for proxy chaining, new arg naming, etc), but now that I know cchurch's trick for smuggling vars in v1, I might backport the chaining stuff to this one as a standalone connection plugin. Let's leave this one open for now, and I'll try to prioritize the v2 work and backport.

sivel commented:
This PR was superseded by the already merged https://github.com/ansible/ansible/pull/11908
